pg-extend/src/pg_datum.rs

```diff
@@ -110,6 +112,75 @@ impl From<i64> for PgDatum {
     }
 }
 
+impl TryFromPgDatum for String {
+    fn try_from(datum: PgDatum) -> Result<Self, &'static str> {
+        use std::ffi::CStr;
+        use std::os::raw::c_char;
+
+        if let Some(datum) = datum.0 {
+            let text_val = datum as *const pg_sys::text;
+
+            let cstr = unsafe {
+                let val: *mut c_char = pg_sys::text_to_cstring(text_val);
+                pg_sys::pfree(text_val as *mut _);
+
+                CStr::from_ptr(val)
+            };
+
+            match cstr.to_str() {
+                Ok(s) => Ok(s.to_owned()),
+                Err(_) => Err("datum was not valid utf8")
+            }
+        } else {
+            Err("datum was NULL")
+        }
+    }
+}
+
+impl From<String> for PgDatum {
+    fn from(value: String) -> Self {
+        use std::ffi::CString;
+        use std::os::raw::c_char;
+
+        let cstr = CString::new(value).expect("This shouldn't fail");
+        let ptr: *const c_char = cstr.as_ptr();
+
+        let text = unsafe { pg_sys::cstring_to_text(ptr) };
+
+        PgDatum(Some(text as Datum))
+    }
+}
+
+impl TryFromPgDatum for CString {
```

Comment:
Can we use this as the basis for all the String conversions? I mean, in the String `TryFromPgDatum` and `From<String> for PgDatum`, you could just use this, and than the unsafe code would all be localized to just this method.